zerver/lib/user_topics.py

```diff
@@ -96,34 +96,37 @@ def set_topic_mutes(
         recipient_id = stream.recipient_id
         assert recipient_id is not None
 
-        add_topic_mute(
+        add_topic_visibility_policy(
             user_profile=user_profile,
             stream_id=stream.id,
             recipient_id=recipient_id,
             topic_name=topic_name,
-            date_muted=date_muted,
+            visibility_policy=UserTopic.MUTED,
+            date_added=date_muted,
         )
 
 
-def add_topic_mute(
+def add_topic_visibility_policy(
     user_profile: UserProfile,
     stream_id: int,
     recipient_id: int,
     topic_name: str,
-    date_muted: Optional[datetime.datetime] = None,
+    visibility_policy: int,
+    date_added: Optional[datetime.datetime] = None,
     ignore_duplicate: bool = False,
 ) -> None:
-    if date_muted is None:
-        date_muted = timezone_now()
+    if date_added is None:
+        date_added = timezone_now()
+
     UserTopic.objects.bulk_create(
         [
             UserTopic(
                 user_profile=user_profile,
                 stream_id=stream_id,
                 recipient_id=recipient_id,
                 topic_name=topic_name,
-                last_updated=date_muted,
-                visibility_policy=UserTopic.MUTED,
+                last_updated=date_added,
+                visibility_policy=visibility_policy,
             ),
```

Comment:
So I'm a bit concerned that we need to rewrite this function. Structurally, if there's an existing row with a given visibility policy, and we're trying to set a different policy, we need to detect that and update the row to the new policy, rather than just trying to make a new row. Usually that's done via a `UserTopic.objects.get_or_create` query or a `select_for_update`.

I think we'll want to rename this to `do_set_topic_visibility_policy` as part of reworking it to achieve our goals here.

781107308da36f3a5ab70408a0f95acb3b2a82e4 has some context; I'm not sure why the commit message doesn't have more context, but @alexmv may be able to help (or maybe you can find the PR containing that commit in case it answers that question).

Reply:
> I think we'll want to rename this to `do_set_topic_visibility_policy` as part of reworking it to achieve our goals here.

You mean renaming `do_set_unmuted_visibility` (in this PR)  to `do_set_topic_visibility_policy`, and replacing `do_mute_topic` and `do_unmute_topic` with it as well.